ethers-contract/src/multicall.rs

```diff
@@ -0,0 +1,204 @@
+extern crate hex;
+
+use ethers_core::{
+    abi::{Detokenize, Function, Token},
+    types::{Address, BlockNumber, NameOrAddress, TransactionRequest, U256},
+};
+use ethers_providers::JsonRpcClient;
+use ethers_signers::{Client, Signer};
+use hex::FromHex;
+use lazy_static::lazy_static;
+
+use std::{collections::HashMap, sync::Arc};
+
+use crate::call::{ContractCall, ContractError};
+use crate::multicall_contract::MulticallContract;
+
+lazy_static! {
+    static ref ADDRESS_BOOK: HashMap<U256, Address> = {
+        let mut m = HashMap::new();
+
+        // mainnet
+        let addr = <[u8; 20]>::from_hex("eefba1e63905ef1d7acba5a8513c70307c1ce441").expect("Decoding failed");
+        m.insert(U256::from(1u8), Address::from(addr));
+
+        // rinkeby
+        let addr = <[u8; 20]>::from_hex("42ad527de7d4e9d9d011ac45b31d8551f8fe9821").expect("Decoding failed");
+        m.insert(U256::from(4u8), Address::from(addr));
+
+        // goerli
+        let addr = <[u8; 20]>::from_hex("77dca2c955b15e9de4dbbcf1246b4b85b651e50e").expect("Decoding failed");
+        m.insert(U256::from(5u8), Address::from(addr));
+
+        // kovan
+        let addr = <[u8; 20]>::from_hex("2cc8688c5f75e365aaeeb4ea8d6a480405a48d2a").expect("Decoding failed");
+        m.insert(U256::from(42u8), Address::from(addr));
+
+        m
+    };
+}
+
+#[derive(Clone)]
+struct Call {
+    target: Address,
+    data: Vec<u8>,
+    function: Function,
+}
+
+pub struct Multicall<P, S> {
+    address: Option<Address>,
+    calls: Vec<Call>,
+    block: Option<BlockNumber>,
+    client: Arc<Client<P, S>>,
+}
+
+impl<P, S> Multicall<P, S> {
+    pub fn new(
+        address: Option<Address>,
+        block: Option<BlockNumber>,
+        client: impl Into<Arc<Client<P, S>>>,
+    ) -> Self {
+        Self {
+            address,
+            calls: vec![],
+            block,
+            client: client.into(),
```

Comment:
Instead of using a client, I think it'd be better if we stored a `MulticallContract`, that would also remove the need for instantiating it inside the `call` method